### PR TITLE
[Rebase & FF] [202405] 2405 integration for supervisor

### PR DESCRIPTION
**Pull request by kuqin12**
## Description

This change integrates the upstream changes from 2405 release tag into supervisor

- [x] Impacts functionality?
  - **Functionality** - Does the change ultimately impact how firmware functions?
  - Examples: Add a new library, publish a new PPI, update an algorithm, ...
- [x] Impacts security?
  - **Security** - Does the change have a direct security impact on an application,
    flow, or firmware?
  - Examples: Crypto algorithm change, buffer overflow fix, parameter
    validation improvement, ...
- [x] Breaking change?
  - **Breaking change** - Will anyone consuming this change experience a break
    in build or boot behavior?
  - Examples: Add a new library class, move a module to a different repo, call
    a function in a new library class in a pre-existing module, ...
- [ ] Includes tests?
  - **Tests** - Does the change include any explicit test code?
  - Examples: Unit tests, integration tests, robot tests, ...
- [ ] Includes documentation?
  - **Documentation** - Does the change contain explicit documentation additions
    outside direct code modifications (and comments)?
  - Examples: Update readme file, add feature readme file, link to documentation
    on an a separate Web page, ...

## How This Was Tested

Change is tested on QEMU Q35 branch.

## Integration Instructions

N/A

### CODE REVIEW

File: MmSupervisorPkg/Core/Handler/Mmi.c [hunk 1]
```diff
@@ -228,15 +283,78 @@ MmiManage (
...
     }
+
+    if (WillReturn) {
+      break;
+    }
   }
 
-  if (SuccessReturn) {
-    Status = EFI_SUCCESS;
+  ASSERT (mMmiManageCallingDepth > 0);
```
Comment (apop5) [inline MmSupervisorPkg/Core/Handler/Mmi.c:292]: PanicLib will catch this even with asserts disabled. Should this get switched over to use it instead?

**Review by apop5 — APPROVED** [review]

**Review by os-d — APPROVED** [review]

File: MmSupervisorPkg/Core/Handler/Mmi.c (hunk 1 above)
Comment (kuqin12) [reply MmSupervisorPkg/Core/Handler/Mmi.c]: Updated.

File: MmSupervisorPkg/Core/Relocate/Relocate.h [hunk 2]
```diff
@@ -739,4 +739,15 @@ ProcessUserSaveStateAccess (
   IN UINT64               Arg3
   );
 
+/**
+  Function to perform post relocation logic before handing back to the IPL.
+
+  TODO: Inspect the page tables. The pages allocated in this SMI may not be protected.
```
Comment (makubacki) [inline MmSupervisorPkg/Core/Relocate/Relocate.h:745]: Might be good to track this todo with an issue in the repo.

**Review by makubacki — APPROVED** [review]

File: MmSupervisorPkg/Core/Relocate/Relocate.h (hunk 2 above)
Comment (kuqin12) [reply MmSupervisorPkg/Core/Relocate/Relocate.h:745]: This should not be there anymore. The implementation was updated to keep the original page table configuration flow.

**Pull request merged**